File: apps/backend/app/models/device.py
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy import ForeignKey, String, DateTime, func, Index
from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID

from app.db.base import Base
from app.models.mixins import TombstoneMixin, UUIDMixin, VersionMixin

if TYPE_CHECKING:
    from app.models.user import User


class Device(Base, UUIDMixin, TombstoneMixin, VersionMixin):
    __tablename__ = "devices"

    __table_args__ = (Index("ix_devices_user_id", "user_id"),)

    user_id: Mapped[UUID] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"))
    name: Mapped[str] = mapped_column(String(50))

    last_seen: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    user: Mapped["User"] = relationship(
        back_populates="devices",
    )

File: apps/frontend/lib/types/Device.ts
import { EntityType } from "./EntityType";

export type DeviceRead = {
  id: string;
  userId: string;
  name: string | null;
  lastSeen: number;
};

export type DeviceWrite = {
  name: string | null;
};

export type Device = {
  id: string;
  userId: string;
  name: string | null;
  lastSeen: string;
  deletedAt: string | null;
  version: number;
  entityType: EntityType.Device;
};
